File: src/shared_libs/zeroui_observability/instrumentation/python/instrumentation.py
"""
Python instrumentation for ZeroUI Observability Layer.

Async/non-blocking telemetry emission using OpenTelemetry OTLP.
"""

import asyncio
import logging
import os
import uuid
from datetime import datetime
from typing import Any, Dict, Optional

try:
    from opentelemetry import trace
    from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
    from opentelemetry.sdk.trace import TracerProvider
    from opentelemetry.sdk.trace.export import BatchSpanProcessor
    from opentelemetry.sdk.resources import Resource
    from opentelemetry.trace import Status, StatusCode
    OTLP_AVAILABLE = True
except ImportError:
    OTLP_AVAILABLE = False
    trace = None  # type: ignore

from ...contracts.event_types import EventType
from ...correlation.trace_context import get_or_create_trace_context, TraceContext
from ...privacy.redaction_enforcer import RedactionEnforcer

logger = logging.getLogger(__name__)


class EventEmitter:
    """
    Event emitter for ZeroUI Observability Layer.

    Emits events using Phase 0 contracts (envelope + payload schemas).
    Applies redaction before emission and propagates trace context.
    """

    def __init__(
        self,
        otlp_endpoint: Optional[str] = None,
        enabled: bool = True,
        component: str = "backend",
        channel: str = "backend",
        version: Optional[str] = None,
    ):
        """
        Initialize event emitter.

        Args:
            otlp_endpoint: OTLP exporter endpoint (defaults to env var)
            enabled: Whether emission is enabled (defaults to feature flag)
            component: Component name
            channel: Channel name (ide, edge_agent, backend, ci, other)
            version: Component version
        """
        self._enabled = enabled and self._check_feature_flag()
        self._component = component
        self._channel = channel
        self._version = version or os.getenv("COMPONENT_VERSION", "unknown")

        # Initialize redaction enforcer
        self._redaction_enforcer = RedactionEnforcer(use_cccs=False)

        # Initialize OTLP if available
        self._otlp_initialized = False
        if OTLP_AVAILABLE and self._enabled:
            self._init_otlp(otlp_endpoint)

    def _check_feature_flag(self) -> bool:
        """Check if observability is enabled via feature flag."""
        return os.getenv("ZEROUI_OBSV_ENABLED", "true").lower() == "true"

    def _init_otlp(self, endpoint: Optional[str] = None) -> None:
        """Initialize OpenTelemetry OTLP exporter."""
        if not OTLP_AVAILABLE:
            logger.warning("OpenTelemetry not available, event emission disabled")
            return

        try:
            endpoint = endpoint or os.getenv("OTLP_EXPORTER_ENDPOINT", "http://localhost:4317")

            # Create resource
            resource = Resource.create({
                "service.name": self._component,
                "service.version": self._version,
                "zeroui.channel": self._channel,
            })

            # Create tracer provider
            tracer_provider = TracerProvider(resource=resource)
            trace.set_tracer_provider(tracer_provider)

            # Create OTLP exporter
            otlp_exporter = OTLPSpanExporter(endpoint=endpoint)

            # Add batch processor
            span_processor = BatchSpanProcessor(otlp_exporter)
            tracer_provider.add_span_processor(span_processor)

            self._otlp_initialized = True
            logger.info(f"OTLP exporter initialized: {endpoint}")
        except Exception as e:
            logger.error(f"Failed to initialize OTLP exporter: {e}", exc_info=True)

    async def emit_event(
        self,
        event_type: EventType,
        payload: Dict[str, Any],
        severity: str = "info",
        trace_ctx: Optional[TraceContext] = None,
    ) -> bool:
        """
        Emit observability event asynchronously.

        Args:
            event_type: Event type (from EventType enum)
            payload: Event payload (must match event type schema)
            severity: Event severity (debug, info, warn, error, critical)
            trace_ctx: Optional trace context (creates new if not provided)

        Returns:
            True if event was emitted, False otherwise
        """
        if not self._enabled:
            return False

        try:
            # Get or create trace context
            if trace_ctx is None:
                trace_ctx = get_or_create_trace_context()

            # Apply redaction to payload
            redaction_result = self._redaction_enforcer.enforce(
                payload,
                compute_fingerprints=True,
            )
            redacted_payload = redaction_result.redacted_payload
            redacted_payload["redaction_applied"] = redaction_result.redaction_applied

            # Create event envelope
            event = self._create_envelope(
                event_type=event_type,
                payload=redacted_payload,
                severity=severity,
                trace_ctx=trace_ctx,
            )

            # Emit via OTLP (async, non-blocking)
            await self._emit_otlp_log(event)

            return True
        except Exception as e:
            logger.error(f"Failed to emit event {event_type.value}: {e}", exc_info=True)
            return False

    def _create_envelope(
        self,
        event_type: EventType,
        payload: Dict[str, Any],
        severity: str,
        trace_ctx: TraceContext,
    ) -> Dict[str, Any]:
        """
        Create event envelope per zero_ui.obsv.event.v1 schema.

        Args:
            event_type: Event type
            payload: Event payload
            severity: Event severity
            trace_ctx: Trace context

        Returns:
            Event envelope dictionary
        """
        return {
            "event_id": f"evt_{uuid.uuid4().hex[:16]}",
            "event_time": datetime.utcnow().isoformat() + "Z",
            "event_type": event_type.value,
            "severity": severity,
            "source": {
                "component": self._component,
                "channel": self._channel,
                "version": self._version,
            },
            "correlation": {
                "trace_id": trace_ctx.trace_id,
                "span_id": trace_ctx.span_id,
            },
            "payload": payload,
        }

    async def _emit_otlp_log(self, event: Dict[str, Any]) -> None:
        """
        Emit event as OTLP log record (async, non-blocking).

        Args:
            event: Event envelope dictionary
        """
        if not self._otlp_initialized or not OTLP_AVAILABLE:
            logger.debug("OTLP not initialized, skipping emission")
            return

        try:
            # Use OpenTelemetry logging API to emit as log record
            # Events are transported as OTLP Logs per architecture
            tracer = trace.get_tracer(__name__)
            with tracer.start_as_current_span("zeroui.obsv.event") as span:
                # Set span attributes from event
                span.set_attribute("event.id", event["event_id"])
                span.set_attribute("event.type", event["event_type"])
                span.set_attribute("event.severity", event["severity"])
                span.set_attribute("zeroui.component", event["source"]["component"])
                span.set_attribute("zeroui.channel", event["source"]["channel"])

                # Log event as structured log (will be exported as OTLP log)
                logger.info(
                    "ZeroUI observability event",
                    extra={
                        "zeroui_event": event,
                        "trace_id": event["correlation"]["trace_id"],
                        "span_id": event["correlation"]["span_id"],
                    },
                )
        except Exception as e:
            logger.error(f"Failed to emit OTLP log: {e}", exc_info=True)

    async def emit_perf_sample(
        self,
        operation: str,
        latency_ms: int,
        trace_ctx: Optional[TraceContext] = None,
        **kwargs: Any,
    ) -> bool:
        """
        Emit perf.sample.v1 event.

        Args:
            operation: Operation name
            latency_ms: Latency in milliseconds
            trace_ctx: Optional trace context
            **kwargs: Additional payload fields (cache_hit, async_path, queue_depth, etc.)

        Returns:
            True if event was emitted
        """
        payload = {
            "operation": operation,
            "latency_ms": latency_ms,
            "component": self._component,
            "channel": self._channel,
            **kwargs,
        }
        return await self.emit_event(
            EventType.PERF_SAMPLE,
            payload,
            severity="info",
            trace_ctx=trace_ctx,
        )

    async def emit_error_captured(
        self,
        error_class: str,
        error_code: str,
        stage: str,
        trace_ctx: Optional[TraceContext] = None,
        **kwargs: Any,
    ) -> bool:
        """
        Emit error.captured.v1 event.

        Args:
            error_class: Error class (data, architecture, prompt, retrieval, memory, tool, orchestration, security, unknown)
            error_code: Error code
            stage: Stage where error occurred
            trace_ctx: Optional trace context
            **kwargs: Additional payload fields (message_fingerprint, input_fingerprint, etc.)

        Returns:
            True if event was emitted
        """
        import hashlib

        # Compute fingerprints for required fields
        message = kwargs.get("message", "")
        message_fingerprint = hashlib.sha256(message.encode("utf-8")).hexdigest() if message else ""

        payload = {
            "error_class": error_class,
            "error_code": error_code,
            "stage": stage,
            "message_fingerprint": message_fingerprint,
            "input_fingerprint": kwargs.get("input_fingerprint", ""),
            "output_fingerprint": kwargs.get("output_fingerprint", ""),
            "internal_state_fingerprint": kwargs.get("internal_state_fingerprint", ""),
            "component": self._component,
            "channel": self._channel,
            **{k: v for k, v in kwargs.items() if k not in ["message"]},
        }
        return await self.emit_event(
            EventType.ERROR_CAPTURED,
            payload,
            severity="error",
            trace_ctx=trace_ctx,
        )


# Global event emitter instance
_event_emitter: Optional[EventEmitter] = None


def get_event_emitter() -> EventEmitter:
    """
    Get global event emitter instance (singleton).

    Returns:
        EventEmitter instance
    """
    global _event_emitter
    if _event_emitter is None:
        _event_emitter = EventEmitter()
    return _event_emitter

File: src/shared_libs/zeroui_observability/instrumentation/typescript/instrumentation.ts
/**
 * TypeScript instrumentation for ZeroUI Observability Layer.
 *
 * Async/non-blocking telemetry emission for VS Code Extension (Tier 1) and Edge Agent (Tier 2).
 * Does not block UI thread.
 */

import { EventType } from '../../contracts/event_types';
import { getOrCreateTraceContext, TraceContext } from '../../correlation/trace_context';

/**
 * Event emitter for ZeroUI Observability Layer.
 *
 * Emits events using Phase 0 contracts (envelope + payload schemas).
 * Applies redaction before emission and propagates trace context.
 * Async/non-blocking - does not block UI thread.
 */
export class EventEmitter {
  private enabled: boolean;
  private component: string;
  private channel: string;
  private version: string;
  private otlpEndpoint: string;

  constructor(
    otlpEndpoint?: string,
    enabled: boolean = true,
    component: string = 'edge_agent',
    channel: string = 'ide',
    version?: string
  ) {
    this.enabled = enabled && this.checkFeatureFlag();
    this.component = component;
    this.channel = channel;
    this.version = version || process.env.COMPONENT_VERSION || 'unknown';
    this.otlpEndpoint = otlpEndpoint || process.env.OTLP_EXPORTER_ENDPOINT || 'http://localhost:4317';
  }

  private checkFeatureFlag(): boolean {
    return (process.env.ZEROUI_OBSV_ENABLED || 'true').toLowerCase() === 'true';
  }

  /**
   * Emit observability event asynchronously (non-blocking).
   *
   * @param eventType Event type
   * @param payload Event payload (must match event type schema)
   * @param severity Event severity (debug, info, warn, error, critical)
   * @param traceCtx Optional trace context (creates new if not provided)
   * @returns Promise resolving to true if event was emitted
   */
  async emitEvent(
    eventType: EventType,
    payload: Record<string, any>,
    severity: string = 'info',
    traceCtx?: TraceContext
  ): Promise<boolean> {
    if (!this.enabled) {
      return false;
    }

    try {
      // Get or create trace context
      const ctx = traceCtx || getOrCreateTraceContext();

      // Apply redaction to payload (basic - full redaction in Phase 0)
      const redactedPayload = this.applyBasicRedaction(payload);
      redactedPayload.redaction_applied = true;

      // Create event envelope
      const event = this.createEnvelope(eventType, redactedPayload, severity, ctx);

      // Emit via OTLP (async, non-blocking)
      await this.emitOtlpLog(event);

      return true;
    } catch (error) {
      console.error(`Failed to emit event ${eventType}:`, error);
      return false;
    }
  }

  private createEnvelope(
    eventType: EventType,
    payload: Record<string, any>,
    severity: string,
    traceCtx: TraceContext
  ): Record<string, any> {
    return {
      event_id: `evt_${crypto.randomUUID().substring(0, 16)}`,
      event_time: new Date().toISOString(),
      event_type: eventType,
      severity: severity,
      source: {
        component: this.component,
        channel: this.channel,
        version: this.version,
      },
      correlation: {
        trace_id: traceCtx.traceId,
        span_id: traceCtx.spanId,
      },
      payload: payload,
    };
  }

  private applyBasicRedaction(payload: Record<string, any>): Record<string, any> {
    // Basic redaction - remove deny-listed fields
    const denyList = ['api_key', 'password', 'secret', 'raw_input', 'raw_output', 'raw_message'];
    const redacted = { ...payload };
    for (const field of denyList) {
      if (field in redacted) {
        delete redacted[field];
      }
    }
    return redacted;
  }

  private async emitOtlpLog(event: Record<string, any>): Promise<void> {
    // Emit via OTLP HTTP endpoint (async, non-blocking)
    // Use fetch with fire-and-forget pattern to avoid blocking UI
    fetch(this.otlpEndpoint + '/v1/logs', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        resourceLogs: [
          {
            resource: {
              attributes: [
                { key: 'service.name', value: { stringValue: this.component } },
                { key: 'service.version', value: { stringValue: this.version } },
                { key: 'zeroui.channel', value: { stringValue: this.channel } },
              ],
            },
            scopeLogs: [
              {
                scope: { name: 'zeroui.observability' },
                logRecords: [
                  {
                    timeUnixNano: Date.now() * 1000000,
                    severityNumber: this.getSeverityNumber(event.severity),
                    severityText: event.severity,
                    body: { stringValue: JSON.stringify(event) },
                    attributes: [
                      { key: 'event.id', value: { stringValue: event.event_id } },
                      { key: 'event.type', value: { stringValue: event.event_type } },
                      { key: 'trace_id', value: { stringValue: event.correlation.trace_id } },
                      { key: 'span_id', value: { stringValue: event.correlation.span_id } },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      }),
    }).catch((error) => {
      // Silently fail to avoid blocking UI
      console.debug('OTLP emission failed (non-blocking):', error);
    });
  }

  private getSeverityNumber(severity: string): number {
    const severityMap: Record<string, number> = {
      debug: 5,
      info: 9,
      warn: 13,
      error: 17,
      critical: 21,
    };
    return severityMap[severity] || 9;
  }

  /**
   * Emit perf.sample.v1 event.
   */
  async emitPerfSample(
    operation: string,
    latencyMs: number,
    traceCtx?: TraceContext,
    extra?: Record<string, any>
  ): Promise<boolean> {
    const payload = {
      operation: operation,
      latency_ms: latencyMs,
      component: this.component,
      channel: this.channel,
      ...extra,
    };
    return await this.emitEvent(EventType.PERF_SAMPLE, payload, 'info', traceCtx);
  }

  /**
   * Emit error.captured.v1 event.
   */
  async emitErrorCaptured(
    errorClass: string,
    errorCode: string,
    stage: string,
    traceCtx?: TraceContext,
    extra?: Record<string, any>
  ): Promise<boolean> {
    // Compute fingerprints (basic - full implementation in Phase 0)
    const message = extra?.message || '';
    const messageFingerprint = message
      ? await this.computeFingerprint(message)
      : '';

    const payload = {
      error_class: errorClass,
      error_code: errorCode,
      stage: stage,
      message_fingerprint: messageFingerprint,
      input_fingerprint: extra?.input_fingerprint || '',
      output_fingerprint: extra?.output_fingerprint || '',
      internal_state_fingerprint: extra?.internal_state_fingerprint || '',
      component: this.component,
      channel: this.channel,
      ...Object.fromEntries(
        Object.entries(extra || {}).filter(([k]) => k !== 'message')
      ),
    };
    return await this.emitEvent(EventType.ERROR_CAPTURED, payload, 'error', traceCtx);
  }

  private async computeFingerprint(content: string): Promise<string> {
    // SHA-256 fingerprint (basic implementation)
    const encoder = new TextEncoder();
    const data = encoder.encode(content);
    const hashBuffer = await crypto.subtle.digest('SHA-256', data);
    const hashArray = Array.from(new Uint8Array(hashBuffer));
    return hashArray.map((b) => b.toString(16).padStart(2, '0')).join('');
  }
}

// Global event emitter instance
let eventEmitter: EventEmitter | null = null;

/**
 * Get global event emitter instance (singleton).
 */
export function getEventEmitter(): EventEmitter {
  if (!eventEmitter) {
    eventEmitter = new EventEmitter();
  }
  return eventEmitter;
}
